Timer - Updated Triggering and Parsing - Fixes #1937

davidofferman opened the pull request:
Hello,
This update fixes some triggering issues with the Timer. This includes fixing triggering such as "five-timer", along with removing the "time" trigger.

This update also adds a few extra triggers. Before the "online" and "for" keyword was only applied to the "timer" trigger, now it's applied to all the `startend` triggers. This allows for queries such as "online alarm" and "countdown for 10 minutes".

I can rollback the triggering even more if you want.

Fixes issue #1937 
Instant Answer: https://duck.co/ia/view/timer

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

jagtalon commented on lib/DDG/Spice/Timer.pm:
```diff
@@ -13,18 +13,50 @@ code_url 'https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/
+    my $trgx = join('|', @triggers);
+
+    # check to make sure the query matches a trigger perfectly
+    if($qry eq '' && in_array($raw, @triggers)) {
```
@davidofferman awesome! I was thinking that maybe instead of implementing an `in_array` subroutine you could use something like this? Tell me what you think! http://stackoverflow.com/a/6475315

jagtalon replied:
I think it's more readable this way

jagtalon commented on lib/DDG/Spice/Timer.pm:
```diff
@@ -13,18 +13,50 @@ code_url 'https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/
+        return '';
+    }
+
+    # makes sure trigger is wrapped with whitespace
```
Give examples in the comment on why we want to make sure that the trigger should be wrapped in whitespace. It's a bit unclear.

jagtalon commented on lib/DDG/Spice/Timer.pm:
```diff
@@ -13,18 +13,50 @@ code_url 'https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/
+        return;
+    }
+
+    $raw =~ s/\s*(online )?($trgx)( online)?( for )?\s*//;
```
Also explain why we want to remove this part in the raw query.

davidofferman commented:
@jagtalon I updated the comments and changed the `in_array` function. This function wasn't needed if I first checked to make sure the raw query trigger is wrapped with white space or is at the start or end of the query.

Hopefully everything is more clear now.

jagtalon commented:
@davidofferman love the comments! :+1:
